cunumeric/deferred.py

```diff
@@ -534,142 +534,233 @@ def _slice_store(k: slice, store: Store, dim: int) -> tuple[slice, Store]:
+
+            key = self._unpack_ellipsis(key, self.ndim)
+
+            # loop through all the keys to check if there
+            # is a single NumPyThunk entry
+            num_arrays = 0
+            transpose_index = 0
+            for dim, k in enumerate(key):
+                if isinstance(k, NumPyThunk):
+                    num_arrays += 1
+                    transpose_index = dim
+
+            # this is the case when there is a single boolean array passed
+            # in this case we transpose original array so that the indx
+            # to which boolean array is passed to goes first
+            # doing this we can avoid going through Realm Copy which should
+            # improve performance
+            if (
+                num_arrays == 1
+                and key[transpose_index].dtype == bool
+                and is_set
+            ):
+                lhs = self
+                key_dim = key[transpose_index].ndim
+                transpose_indices = tuple(
+                    (transpose_index + i) for i in range(0, key_dim)
+                )
+                transpose_indices += tuple(
+                    i for i in range(0, transpose_index)
+                )
+                transpose_indices += tuple(
+                    i for i in range(transpose_index + key_dim, lhs.ndim)
+                )
+
+                new_key = tuple(key[i] for i in range(0, transpose_index))
+                new_key += tuple(
+                    key[i] for i in range(transpose_index + key_dim, len(key))
+                )
+                lhs = lhs.transpose(transpose_indices)
+
+                # transform original array for all other keys in the tuple
+                if len(new_key) > 0:
+                    shift = 0
+                    store = lhs.base
+                    for dim, k in enumerate(new_key):
+                        if np.isscalar(k):
+                            if k < 0:  # type: ignore [operator]
+                                k += store.shape[dim + key_dim + shift]
+                            store = store.project(dim + key_dim + shift, k)
+                            shift -= 1
+                        elif k is np.newaxis:
+                            store = store.promote(dim + key_dim + shift, 1)
+                        elif isinstance(k, slice):
+                            k, store = self._slice_store(
+                                k, store, dim + key_dim + shift
+                            )
+                        else:
+                            raise TypeError(
+                                "Unsupported entry type passed to advanced ",
+                                "indexing operation",
+                            )
+                    lhs = DeferredArray(self.runtime, store, self.dtype)
+
+                return True, lhs, key[transpose_index]
+
+            # this is a general advanced indexing case
+            else:
+                return False, self, key
```

Comment:
I think you're missing an `_unpack_ellipsis` here (you don't need to add it if you change the code to use `bool_key` below).

Reply:
Nevermind, I missed it right after the `assert isinstance(key, tuple)`.